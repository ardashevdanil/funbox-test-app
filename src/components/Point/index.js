import PropTypes from 'prop-types';
import React from 'react';

class Point extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.pointRef = React.createRef();
    this.getCoords = this.getCoords.bind(this);
    this.handleDeleteClick = this.handleDeleteClick.bind(this);
    this.handleMouseDown = this.handleMouseDown.bind(this);
  }

  getCoords() {
    return this.pointRef.current.getBoundingClientRect();
  }
  
  handleDeleteClick(e) {
    e.stopPropagation();
  	this.props.onDeleteClick(this.props.id);
  }

  handleMouseDown(e) {
    e.preventDefault();
    this.setState({style: {background: '#8fffff'}});

    document.onmousemove = (e) => {
      this.props.setPointPosition(
        e.pageY,
        this.props.id
      );
    }

    document.onmouseup = () => {
      document.onmousemove = null;
      document.onmouseup = null;
      this.setState({style: {}});
    }
  }

  componentDidUpdate(prevProps) {
    /*
      setting the point position after updating the list of points
      or scrolling the parent component
    */
    if (
      !this.props.position ||
      this.props.id !== prevProps.id ||
      this.props.position !== prevProps.position ||
      this.props.scrollTop !== prevProps.scrollTop
    ) {
      const nodeHeight = this.pointRef.current.offsetHeight;
      const top = this.getCoords().top + nodeHeight/2;

      this.props.setPointPosition(top, this.props.id);
    }
  }

  render() {
    return (
      <div
        className="point"
        ref={this.pointRef}
        onMouseDown={this.handleMouseDown}
        onDragStart={ () => false }
        style={this.state.style}
      >
        {this.props.value}
        <span
          onMouseDown={this.handleDeleteClick}
          className="delete-button"
        >
          X
        </span>
      </div>
    );
  }
}

Point.propTypes = {
  scrollTop: PropTypes.number.isRequired,
  position: PropTypes.number,
  value: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  id: PropTypes.number.isRequired,
  onDeleteClick: PropTypes.func.isRequired,
  setPointPosition: PropTypes.func.isRequired,
}

export default Point;