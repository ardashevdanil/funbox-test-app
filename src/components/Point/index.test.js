import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';

import Point from './';

Enzyme.configure({ adapter: new Adapter() });

describe('test Point', () => {
  const onDeleteClick = jest.fn();
  const setPointPosition = jest.fn();
	const component = Enzyme.mount(
    <Point 
      position={42}
      value='point'
      key={1}
      id={1}
      onDeleteClick={onDeleteClick}
      setPointPosition={setPointPosition}
      scrollTop={0}
    />
  );

  test('renders correctly', () => {
    expect(component.find('div').at(0).text()).toEqual('pointX');
    expect(component.debug()).toMatchSnapshot();
  });

  test('calls onDeleteClick after click on the delete button', () => {
    component.find('span').simulate('mousedown');
    expect(onDeleteClick.mock.calls.length).toBe(1);    
    expect(onDeleteClick.mock.calls[0][0]).toBe(1);    
  });

  test('calls setPointPosition after mouse down and mouse move', () => {
    const mousemoveEvent = new Event('mousemove');

    mousemoveEvent.pageY = 20;
    component.simulate('mousedown');
    document.dispatchEvent(mousemoveEvent);
    expect(setPointPosition.mock.calls.length).toBe(1);    
    expect(setPointPosition.mock.calls[0]).toEqual([20, 1]);    
  });

  test('calls setPointPosition after updating', () => {
    component.setProps({position: undefined});
    component.update();
    component.setProps({id: 10});
    component.update();
    component.setProps({position: 10});
    component.update();
    component.setProps({scrollTop: 10});
    component.update();
    expect(setPointPosition.mock.calls.length).toBe(5);    
    expect(setPointPosition.mock.calls[1][1]).toEqual(1);    
    expect(setPointPosition.mock.calls[2][1]).toEqual(10);    
  });
});
