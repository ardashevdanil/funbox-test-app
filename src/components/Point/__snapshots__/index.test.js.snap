// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Point renders correctly 1`] = `
"<Point position={42} value=\\"point\\" id={1} onDeleteClick={[Function: mockConstructor]} setPointPosition={[Function: mockConstructor]} scrollTop={0}>
  <div className=\\"point\\" onMouseDown={[Function: bound handleMouseDown]} onDragStart={[Function: onDragStart]} style={[undefined]}>
    point
    <span onMouseDown={[Function: bound handleDeleteClick]} className=\\"delete-button\\">
      X
    </span>
  </div>
</Point>"
`;
